### PR TITLE
[sui framework] use drand round for time more fine-grained than an epoch

### PR DESCRIPTION
**Pull request by sblackshear**
Sui has epoch time (~24 hours) available to Move programmers, but nothing more fine-grained. This is probably the most frequently cited pain point by Sui developers. https://github.com/MystenLabs/sui/issues/226 will eventually solve this, but this is not expected for quite some time.
    
This PR provides an intermediate solution by moving @benr-ml's clever `drand` module into the Sui Framework recording the latest drand round in a shared object. This allows developers to:
    - Use the drand round as a more fine-grained counter than epoch
    - Use drand's round time to derive a Unix timestamp
    
This only works well for proving that a particular time happened in the past, but it is much better than what we have today.
    
In order for this to be useful, the drand shared object must be updated fairly frequently. I suggest that we modify the faucet to send a drand-updating tx every time it gets a request + more than 30 seconds have passed since the last update. Copying @666lcz and @longbowlu for advice on whether this is a good idea + how to do it if so.

### CODE REVIEW

File: crates/sui-framework/sources/crypto/drand.move [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    public fun chain(): vector<u8> {
+        CHAIN
+    }
+
+    /// Return the public key of the drand instance used by this module
+    public fun public_key(): vector<u8> {
+        DRAND_PK
+    }
+
+    /// Returnt he period of the drand instance used by this module
```
Comment (benr-ml) [inline crates/sui-framework/sources/crypto/drand.move:158]: typo (Return the)

File: crates/sui-framework/sources/crypto/drand.move [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    public fun sig(drand: &Drand): vector<u8> {
+        drand.previous_sig
+    }
+
+    /// Return the most recent time we have seen in seconds.
+    /// This time is coarse-grained--it will only advance by increments of PERIOD (30 seconds).
+    /// Note that this is not an up-to-date time unless `drand` has been updated recently
+    /// --it should only be used as a proof that this particular time occurred in the past.
+    public fun unix_time_seconds(drand: &Drand): u64 {
+        GENESIS + (PERIOD * drand.previous_round)
```
Comment (benr-ml) [inline crates/sui-framework/sources/crypto/drand.move:85]: GENESIS + (PERIOD * (drand.previous_round -1))

Genesis = round 1

File: crates/sui-framework/sources/crypto/drand.move [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    /// Return the most recent Unix time we have seen in seconds, according to drand
+    /// This time is coarse-grained--it will only advance by increments of PERIOD (30 seconds).
+    /// Note that this is not an up-to-date time unless `drand` has been updated recently
+    /// --it should only be used as a proof that this particular time occurred in the past.
+    /// Finally, note that although drand is supposed to produce a new round every 30 seconds
+    /// and always has since GENESIS, it is possible for the gaps between rounds to be > or < 30 seconds.
+    /// Thus, this function should only be used as an approximation of Unix time for non-production use-cases
+    /// --for anything security-critical, please use either Sui epoch time or the forthcoming https://github.com/MystenLabs/sui/pull/6490
```
Comment (kchalkias) [inline crates/sui-framework/sources/crypto/drand.move:87]: replace forthcoming link with https://github.com/MystenLabs/sui/issues/226

File: crates/sui-framework/sources/crypto/drand.move [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    /// Note that a return value of `false` does *not* necesarily mean that `unix_timestamp` is in
+    /// in the future--`drand` might just be out-of-date.
+    public fun timestamp_has_passed(drand: &Drand, unix_timestamp: u64): bool {
+        unix_timestamp < approx_unix_time_seconds(drand)
+    }
+
+    // functions that leverage drand without requiring access to the shared object
+
+    /// Check that a given epoch time has passed by verifying a drand signature from a later time.
+    /// round must be at least (epoch_time - GENESIS) / PERIOD + 1).
```
Comment (kchalkias) [inline crates/sui-framework/sources/crypto/drand.move:102]: due to Genesis = round 1, we need to update this comment + logic here with `round - 2`

File: crates/sui-framework/sources/crypto/drand.move [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    /// Return the most recent signature we have seen
+    public fun sig(drand: &Drand): vector<u8> {
+        drand.previous_sig
+    }
+
+    /// Return the most recent Unix time we have seen in seconds, according to drand
+    /// This time is coarse-grained--it will only advance by increments of PERIOD (30 seconds).
+    /// Note that this is not an up-to-date time unless `drand` has been updated recently
```
Comment (kchalkias) [inline crates/sui-framework/sources/crypto/drand.move:82]: I'd add a `WARNING` with uppercase here

**Review by kchalkias — COMMENTED** [review]
We should be very careful how to advertise this time reference module at `std`.
To be clear:
- I'm convinced that for devnet/testnet a drand time-stamping service might be tolerable, but definitely not as an official proposal for mainnet critical applications.
- Drand network works at best effort to maintain the 30sec period, but it shouldn't be considered a forever guarantee. Indeed, according to Drand's lead engineer statement/analysis (@AnomalRoil), until today it performs stable enough and it's the only reason we can tolerate it for non-criticial, non-mainnet dapps.
- We should highlight that this module might not stay at `std` when `mainnet`. It's an external service that we don't control, and we pin an external `chainID` and `PublicKey` that is not generally considered a best practice, especially while Drand will evolve down the road and the current settings might be deprecated or one day be dropped. Of course, at `mainnet` anyone can publish it as a public module.
- This should currently be considered a helper (to be std deprecated) module with the aim to temporarily have a ~30secs time reference, until there is an official Sui solution.
- Note that any user of this module accepts the fact that they trust the drand committee (and all committees of the past), on a) the random beacon secrecy (no collusion) b) in our use case as a time reference, regarding the 30sec interval.

**Comment by sblackshear** [comment]
> We should be very careful how to advertise this time reference module at `std`. To be clear:
> 
> * I'm convinced that for devnet/testnet a drand time-stamping service might be tolerable, but definitely not as an official proposal for mainnet critical applications.
> * Drand network works at best effort to maintain the 30sec period, but it shouldn't be considered a forever guarantee. Indeed, according to Drand's lead engineer statement/analysis (@AnomalRoil), until today it performs stable enough and it's the only reason we can tolerate it for non-criticial, non-mainnet dapps.
> * We should highlight that this module might not stay at `std` when `mainnet`. It's an external service that we don't control, and we pin an external `chainID` and `PublicKey` that is not generally considered a best practice, especially while Drand will evolve down the road and the current settings might be deprecated or one day be dropped. Of course, at `mainnet` anyone can publish it as a public module.
> * This should currently be considered a helper (to be std deprecated) module with the aim to temporarily have a ~30secs time reference, until there is an official Sui solution.
> * Note that any user of this module accepts the fact that they trust the drand committee (and all Drand committees of the past), on a) the shared secret's secrecy (no collusion) b) in our use case as a time reference, regarding the 30sec interval.
> * In theory, the same timestamp service could be implemented using a different (any secure) Oracle provider or just a basic timestamping service operated by some entity (i.e., Sui foundation); using Drand is only for temp convenience.

Agree with all of these points. I do think a `drand` module in the Sui Framework might make sense (if we expect drand usage to be common), but we should definitely set the expectation that the Unix timestamp functionality will be eliminated. I have tweaked the comments + function name of the Unix-timestamp function according to your feedback; let me know if does not send the right message.

File: crates/sui-framework/sources/crypto/drand.move [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    public fun jump(drand: &mut Drand, sig: vector<u8>, previous_sig: vector<u8>, round: u64) {
+        assert!(round > drand.previous_round, ERoundAlreadyPassed);
+        verify_drand_signature(sig, previous_sig, round);
+        drand.previous_sig = sig;
+        drand.previous_round = round;
+    }
+
+    /// Return the most recent round we have seen
+    public fun round(drand: &Drand): u64 {
+        drand.previous_round
```
Comment (kchalkias) [inline crates/sui-framework/sources/crypto/drand.move:75]: a comment maybe for @benr-ml that naming it `previous_round` might be confusing, especially when after we register it, someone (before the next round being published) tries to get the current round seed. In theory it's "most recent registered round". We'll fix wording here in an upcoming PR.

File: crates/sui-framework/sources/crypto/drand.move [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    /// Return the most recent signature we have seen
+    public fun sig(drand: &Drand): vector<u8> {
+        drand.previous_sig
+    }
+
+    /// Return the most recent Unix time we have seen in seconds, according to drand.
+    /// This time is coarse-grained--it will only advance by increments of PERIOD (30 seconds).
+    /// *WARNING*: Although drand is supposed to produce a new round every 30 seconds
+    /// and always has since GENESIS, it is possible for the gaps between rounds to be > or < 30 seconds, and future.
+    /// versions of the protocol may change PERIOD.
```
Comment (kchalkias) [inline crates/sui-framework/sources/crypto/drand.move:87]: *V* uppercase

**Review by kchalkias — APPROVED** [review]
Awesome! Very useful for testnet DeFi and time-lock based dapps.

**Review by benr-ml — APPROVED** [review]

**Comment by 666lcz** [comment]
>  I suggest that we modify the faucet to send a drand-updating tx every time it gets a request + more than 30 seconds have passed since the last update. Copying @666lcz and @longbowlu for advice on whether this is a good idea + how to do it if so.

Sorry for the late reply @sblackshear. This is essentially a cron job running every 30 seconds, and therefore I would prefer to implement this as a separate service that's independent from the faucet. Happy to take on this task if you need any help.

**Comment by longbowlu** [comment]
I'm not familiar with drand so this may be a very dumb question: 
If we rely on an (perhaps not highly reliable) off-chain service to update the drand round and assume each drand round is 30 second, why don't we just have that off-chain service to provide the unix time? like report the unixtime every 30 sec or 10 sec? 

Btw, if we want to experiment with this before investing resources in a dedicated cron job, we can begin with aws lamba.

**Comment by AnomalRoil** [comment]
FYI, drand does provide one guarantee related to time, which is that a given round is never happening earlier than its expected time. 

So if we have a period of 30 seconds and start at round 10, the guarantee drand provides is that the round 15 won't happen until at least 5*30 seconds have lapsed. 

It's true that the drand network might be degraded at some point (it never happened in the 2.5+ years it has been running so far) and then a round might be delayed. (But then the network would catch up at twice its period.)

**Pull request closed without merging**